nixos/stratis: initrd support for stratis root volumes

mberndt123 opened the pull request:
Hey,

I added initrd support for booting from a [Stratis](https://stratis-storage.github.io/) volume. It's mostly based on the way that it's done in Stratis' dracut module, except that I've decided not to use the [stratis_setup_generator](https://github.com/stratis-storage/stratisd/blob/master/src/bin/generators/stratis_setup_generator.rs) and instead just added a systemd service configuration with the required STRATIS_ROOTFS_UUID env variable. 
###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

mberndt123 committed:
stratisd: split out initrd support

mberndt123 committed:
nixos/stratis: enable booting from stratis volume

ElvishJerricco commented on nixos/modules/system/boot/stratisroot.nix:
```diff
@@ -0,0 +1,76 @@
+        assertion = lib.strings.hasPrefix "/dev/stratis/" config.fileSystems."/".device;
+        message = "root file system must live in /dev/stratis";
+      }
+      {
+        assertion = config.boot.plymouth.enable;
+        message = "plymouth must be enabled for stratis pool password prompt";
+      }
```
This assertion is probably not true. If it's using `systemd-ask-password`, then any ask-password agent will work, such as the console one that systemd stage 1 has by default. Can you verify whether or not that's correct?

NickCao replied:
`systemd-ask-password` is only called in `stratis-fstab-setup`, which is not being used in this PR?

ElvishJerricco commented on nixos/modules/system/boot/stratisroot.nix:
```diff
@@ -0,0 +1,76 @@
+            conflicts = [ "shutdown.target" ];
+            onFailure = [ "emergency.target" ];
+            unitConfig.OnFailureJobMode = "isolate";
+            wants = [ "stratisd-min.service" "plymouth-start.service" "stratis-clevis-setup.service" ];
```
Shouldn't have `plymouth-start.service` in here if plymouth isn't actually required.

What is `stratis-clevis-setup.service`?

NickCao replied:
It is a generated unit, by `stratis-clevis-setup-generator`. mostly the same as `stratis-clevis-setup.service`, but for clevis encrypted volumes. We can safely omit it if not supporting clevis for now.

mberndt123 replied:
For now, Plymouth is required because the stratis-rootfs-setup calls it directly.

I've removed the stratis-clevis-setup.service file

ElvishJerricco commented on nixos/modules/system/boot/stratisroot.nix:
```diff
@@ -0,0 +1,76 @@
+            unitConfig.OnFailureJobMode = "isolate";
+            wants = [ "stratisd-min.service" "plymouth-start.service" "stratis-clevis-setup.service" ];
+            wantedBy = [ "initrd.target" ];
+            after = [ "paths.target" "plymouth-start.service" "stratisd-min.service" ];
```
Ditto on `plymouth-start.service`

mberndt123 replied:
For the time being this is necessary.

ElvishJerricco commented on nixos/modules/system/boot/stratisroot.nix:
```diff
@@ -0,0 +1,76 @@
+          stratis-setup = {
+            description = "setup for Stratis root filesystem";
+            unitConfig.DefaultDependencies = "no";
+            conflicts = [ "shutdown.target" ];
```
Should probably include `initrd-switch-root.target`, as well as putting `shutdown.target` and `initrd-switch-root.target` in the `before` clause. It's not really necessary in this case but it's good practice. Basically, it's how you ensure a unit is properly stopped before reaching `initrd-switch-root.target` or `shutdown.target`. Again, doesn't really matter here because it's a oneshot with no `ExecStop`, but it's good practice. e.g. Someone might add an override to the unit that gives it an `ExecStop` for some reason.

mberndt123 replied:
OK, I've added that

ElvishJerricco reviewed (changes requested):
It'd be great to get something in `nixos/tests/` for this.

NickCao commented:
Still I prefer using the generator, that would give us the opportunity to use stratis in the initrd without being restricted to the rootfs.

mberndt123 committed:
address review comments

mberndt123 commented:
@NickCao ,

I'm afraid I don't understand your comment regarding the generator. All it does is read the `stratis.rootfs.pool_uuid` kernel parameter and generate a service unit with an `Environment=STRATIS_ROOTFS_UUID=…` line from that. The reason they do it that way is that you can use the same initrd to boot from different stratis volumes just by changing the kernel command line. But in NixOS that is impossible anyway because there is no `root=…` parameter on the kernel command line, so being able to configure the pool UUID on the kernel command line is pointless.

NickCao commented:
Ah I confuse the fstab generator with the root setup ones, now I see that your implementation is reasonable. A bit busy to review this these days, would you mind adding a test for this first? A side note: why split the initrd related files to a separate output? These are merely text files and should be pretty small.

mberndt123 commented:
@ElvishJerricco,

I've added an installer test like you suggested on Matrix, but it doesn't run. The last couple of lines in the log are:

```
machine # copying path '/nix/store/yvis74nnab0s8pcbyyb0k6glbbjm7viz-glibc-2.37-8-bin' to 'local'...
machine # copying path '/nix/store/d2i1ikxm4ydlq41mbm67w2yikyiswrf9-gmp-6.2.1' to 'local'...
machine # copying path '/nix/store/mln3mxagq1q7x2yg8zcxmmalhambs991-getent-glibc-2.37-8' to 'local'...
machine # copying path '/nix/store/bhhb91angi23wsr42apz1in1n9b9ilg3-glibc-2.37-8-dev' to 'local'...
machine # error: unable to download 'https://ftpmirror.gnu.org/automake/automake-1.16.5.tar.xz': Couldn't resolve host name (6)
machine # error: filesystem error: cannot rename: Invalid cross-device link [/mnt/nix/store/3yar2pnvz7ll79z3jlzx09qnhrsi7zj5-automake-1.16.5.tar.xz.drv.chroot/nix/store/gmaq49vzfrkvr714y4fhfxv100ijihin-automake-1.16.5.tar.xz] [/ni>
machine: output: 
Test "Perform the installation" failed with error: "command `nixos-install < /dev/null >&2` failed (exit code 1)"
cleanup
kill machine (pid 5)
machine # qemu-kvm: terminating on signal 15 from pid 4 (/nix/store/6jx4nc20v3cfphlbypqid30q41i5vh5x-python3-3.10.11/bin/python3.10)
(finished: cleanup, in 0.13 seconds)
```

I have no idea what's going wrong here 🤷🏻‍♂️

mberndt123 commented:
My approach of tweaking the configuration (with the `amendConfig` snippet) seems to be incompatible with how this test works. So I'll probably need to bite the bullet and adapt the `nixos-generate-config` script to generate a suitable configuration to make this work.

mberndt123 commented:
@NickCao, @ElvishJerricco,

There is now an installer test for this. Since Stratis can be used without encryption, Plymouth is actually not strictly required, so I've decided to remove the Plymouth requirement. And since the `nixos-generate-config` script generates device names like `/dev/disk/by-uuid/2862e89f-b2d6-4a4b-83d4-61c4a5aecc3c` rather than `/dev/stratis/foo/bar` and that works just as well, I've decided to remove that assertion too.

mberndt123 committed:
add installer test for stratis root fs

ElvishJerricco commented:
@mberndt123 I'm not a fan of this `amendConfig` concept, since we have `extraConfig` already. I can see why it's done; we don't know at eval time what the UUID is going to be. But I think there's a couple of alternatives that would be better.

1. You could put in `extraConfig` something like `boot.stratis.rootPoolUuid = builtins.readFile ./uuid.txt;` and then create `/mnt/etc/nixos/uuid.txt` in `createPartitions`.
2. If stratis will allow you to manually select a UUID when creating the pool, you could just hard code one.

I would prefer the second, but I don't know if stratis will allow you to choose the UUID.

That said, this does raise an interesting point. Shouldn't the `boot.stratis.rootPoolUuid` option be set in `hardware-configuration.nix` and generated by `nixos-generate-config`?

ElvishJerricco commented on nixos/modules/system/boot/stratisroot.nix:
```diff
@@ -0,0 +1,68 @@
+      example = "04c68063-90a5-4235-b9dd-6180098a20d9";
+    };
+  };
+  config = {
```
This will need to be behind a `mkIf` depending on if stratis is used in the config.

mberndt123 replied:
done

mberndt123 committed:
address code review issues

mberndt123 commented on nixos/tests/installer.nix:
```diff
@@ -1017,4 +1021,54 @@ in {
+    '';
+    bootLoader = "systemd-boot";
+    extraInstallerConfig = { modulesPath, ...}: {
+      imports = [ (modulesPath + "/tasks/stratis.nix") ];
```
done

mberndt123 commented on nixos/tests/installer.nix:
```diff
@@ -1017,4 +1021,54 @@ in {
+          uuid = pool_line[index - 32: index + 4]
+          amendments = f"""{{ modulesPath, ... }}: {{
+            imports = [
+              (modulesPath + "/system/boot/stratisroot.nix")
```
done

mberndt123 commented:
@ElvishJerricco,

You raise an interesting point with the `nixos-generate-config` thing. I agree that it should generate appropriate stratis configuration, but I think that setting `boot.stratis.rootPoolUuid` is actually not enough. Every fs can potentially sit on a stratis pool, and every fs that has `neededForBoot` set to `true` should be mounted in the initramfs. So I think it would actually make much more sense to add a `stratis.poolUuid` option in `fileSystems` and then do all the other configuration based on that. Makes sense? Wdyt?

mberndt123 committed:
make nixos-generate-config generate stratis pool UUIDs

mberndt123 committed:
improve stratis initrd support

it is now possible to supply a stratis pool uuid
for every filesystem, and if that filesystem
is required for boot, the relevant pool will be
started in the initramfs.

mberndt123 committed:
Merge remote-tracking branch 'upstream/master' into mberndt123/stratis-rootfs

mberndt123 commented:
I've now changed this so that the stratis pool UUID can be configured for every individual file system, and if any of them are required for booting, then initramfs support will be enabled automatically.

mberndt123 committed:
minor refactoring

mberndt123 committed:
remove unneeded import

mberndt123 commented on nixos/modules/system/boot/stratisroot.nix:
```diff
@@ -0,0 +1,64 @@
+                unitConfig.DefaultDependencies = "no";
+                conflicts = [ "shutdown.target" "initrd-switch-root.target" ];
+                onFailure = [ "emergency.target" ];
+                unitConfig.OnFailureJobMode = "isolate";
```
No particular reason, I just modeled it after this:
https://github.com/stratis-storage/stratisd/blob/7d0125be75a68d25329a6e72db4b581fadee536b/src/bin/generators/stratis_setup_generator.rs#L44
I can change it to `replace-irreversibly` if you want.

mberndt123 commented on nixos/modules/tasks/filesystems.nix:
```diff
@@ -36,6 +36,15 @@ let
         description = lib.mdDoc "Location of the mounted file system.";
       };
 
+      stratis.poolUuid = lib.mkOption {
+        type = types.uniq (types.nullOr types.str);
+        description = lib.mdDoc ''
+          UUID of the stratis pool that the fs is located in
+        '';
+        example = "04c68063-90a5-4235-b9dd-6180098a20d9";
+        default = null;
+      };
+
```
Yes, sure. But I don't see how that creates a problem. Due to how `attrsets.mapAttrs'` works, this is automatically deduplicated, so only one stage 1 service unit will be generated per pool.

I could put this information somewhere else, I just found it convenient this way.

ElvishJerricco replied:
On second thought, I think this makes the most sense. Associating it with the FS gives the most useful information.

mberndt123 commented on nixos/tests/installer.nix:
```diff
@@ -989,4 +989,40 @@ in {
+      machine.succeed(
+        "sgdisk --zap-all /dev/vda",
+        "sgdisk --new=1:0:+100M --typecode=0:ef00 /dev/vda", # /boot
+        "sgdisk --new=2:0:+1G --typecode=0:8200 /dev/vda", # swap
```
No, Stratis is unrelated to LVM and a simpler design. There are no logical volumes like in LVM, so you can't swap directly on a stratis pool. You could create a swap file on top of a stratis fs, but I wouldn't see the point.

mberndt123 commented on nixos/tests/installer.nix:
```diff
@@ -989,4 +989,40 @@ in {
+    '';
+    bootLoader = "systemd-boot";
+    extraInstallerConfig = { modulesPath, ...}: {
+      imports = [ (modulesPath + "/tasks/stratis.nix") ];
```
I've removed it.

ElvishJerricco commented:
Thanks!
